src/server/auth/server/api_server.go

```diff
@@ -188,6 +192,33 @@ func NewAuthServer(pachdAddress string, etcdAddress string, etcdPrefix string) (
 	}
 	go s.getPachClient() // initialize connection to Pachd
 	go s.watchAdmins(path.Join(etcdPrefix, adminsPrefix))
+
+	// Retrieve the PPS master token, or generate it and put it in etcd.
+	// TODO This is a hack. It avoids the need to return superuser tokens from
+	// GetAuthToken (essentially, PPS and Auth communicate through etcd instead of
+	// an API) but we should define an internal API and use that instead.
+	var tokenProto types.StringValue // will contain PPS token
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+	if _, err = col.NewSTM(ctx, etcdClient, func(stm col.STM) error {
+		superUserTokenCol := col.NewCollection(etcdClient, ppsconsts.PPSTokenKey, nil, &types.StringValue{}, nil).ReadWrite(stm)
+		err := superUserTokenCol.Get("", &tokenProto)
+		if err == nil {
+			return nil
+		}
+		if col.IsErrNotFound(err) {
+			// no existing token yet -- generate token
+			token := uuid.NewWithoutDashes()
+			tokenProto.Value = token
+			if err := superUserTokenCol.Create("", &tokenProto); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	s.ppsToken = tokenProto.Value
```

Comment:
Could this be done in a goro like the other startup calls? Or does construction need to block on it?

Reply:
I think so—I moved it into a separate goro, and instead of returning an error, the goro panics if it can't initialize the PPS token within 60s. Let me know what you think